### ISSUE TITLE
Modify the packet frame structure to achieve switch information acquisition

### ISSUE THREAD

**Issue by La-Joya** [opennetworkinglab/ngsdn-tutorial#98]
Hello, I encountered some problems when I was learning to use this environment, I hope to get your help and answers.
I made some small changes to the P4 source file, adding an INT structure between IPv6 and TCP, hoping to use it to store the timestamp and queue length of the packet when it entered the switch:
```
typedef bit<32> timestamp_t;
typedef bit<32> qdepth_t;
const bit<8> IP_PROTO_INT    = 150;
......
/*Header*/
header switch_t {
    timestamp_t  qtimestamp;
    qdepth_t    qdepth;
}
......
struct parsed_headers_t {
    cpu_out_header_t cpu_out;
    cpu_in_header_t cpu_in;
    ethernet_t ethernet;
    ipv4_t ipv4;
    ipv6_t ipv6;

    switch_t swtraces;

    srv6h_t srv6h;
    srv6_list_t[SRV6_MAX_HOPS] srv6_list;
    tcp_t tcp;
    udp_t udp;
    icmp_t icmp;
    icmpv6_t icmpv6;
    ndp_t ndp;
}
......
/*Parser*/
    state parse_ipv6 {
        packet.extract(hdr.ipv6);
        local_metadata.ip_proto = hdr.ipv6.next_hdr;
        transition select(hdr.ipv6.next_hdr) {
            IP_PROTO_INT: parse_int;
            IP_PROTO_TCP: parse_tcp;
            IP_PROTO_UDP: parse_udp;
            IP_PROTO_ICMPV6: parse_icmpv6;
            IP_PROTO_SRV6: parse_srv6;
            default: accept;
        }
    }

    state parse_int {
        packet.extract(hdr.swtraces);
        transition parse_tcp;
    } 
    state parse_tcp{

    }
......
/*EgressPipeImpl*/
        if (hdr.ipv6.next_hdr == IP_PROTO_INT) {
            hdr.swtraces.setValid();
            hdr.swtraces.qtimestamp = (timestamp_t)standard_metadata.enq_timestamp;
            hdr.swtraces.qdepth = (qdepth_t)standard_metadata.deq_qdepth;
        }
```
The main operation is to judge the IPv6 next header in the Egress stage. If it is a custom INT header (150), the INT field will be activated and added such as timestamp and queue length.
In order to judge whether my changes are effective, I use the send.py and receive.py in the P4lang exercise to send and receive packets, and use the scapy library：
```
.......
pkt = pkt /IPv6(dst=addr,nh=150) /SwitchTrace() /TCP(dport=1234, sport=random.randint(49152,65535)) /sys.argv[2]
.......
```
But the receiver can't receive any data packets. 
I don't know how to debug. After all, the P4 program can be compiled and passed.
（Using the scapy library to send ordinary IPv6 packets can successfully send and receive.）

PS:Can I use wireshark to capture packets in mininet's docker?

**Comment by ederollora** [comment]
Hi,

TL;DR: Place your INT header after L4 header (TCP or UDP).

Here my two penny worth opinion. If you are sure that egress switch is sending the packet to the end-host then the issue, to me, seems to be the location of the INT header. If you check the original INT spec, you'll reliase that one the of many options to locate the INT header is "over TCP or UDP". This is very important because if you place INT between IPv4/v6 and TCP or UDP, the end-host will "confuse" your L4 heaader. It will not know what 150 in IP proto means and maybe drop the packet. See the example below.

![image](https://user-images.githubusercontent.com/3457431/164006494-e22c1846-2bb6-4c20-9525-b9b258963bc5.png)

If you left TCP or UDP code (6 or 17) in IP proto, the end-host will think that the first 20 or 8 bytes after IP belong to TCP or UDP, respectively. In reality the 20 or 8 bytes will first belong to INT and then to whatever comes after all that. 

Hope it helps,

**Comment by ederollora** [comment]
And finally, yes, you can sniff the traffic but I could never live sniff it. I remember modifying a script for my own purpose. I had to run tshark, save in a pcap and then open with a Wireshark. Check if [this script](https://github.com/opennetworkinglab/ngsdn-tutorial/blob/fbb8b03c2c6c4cd63837e9864ea5c52477404261/util/mn-pcap) already in the repo helps with it.

Cheers,

**Comment by La-Joya** [comment]
I just tried the INT structure after TCP, but it still didn't work.
![image](https://user-images.githubusercontent.com/74955331/164014856-9984dc94-115e-4e75-b7dd-5575903ecb6a.png)
```
pkt = pkt /IPv6(dst=addr,nh=6) /TCP(dport=1234, sport=random.randint(49152,65535)) /SwitchTrace() /sys.argv[2];
```
I chose to put int before the TCP. Because in the int specification, an int example is placed before the TCP after IP.
And in P4LANG exercises, MRI experiment (simple version int) is also place the INT before the TCP.

**Comment by La-Joya** [comment]
Hi，

Of course, I also modified the P4 program accordingly.
```
        if (hdr.ipv6.next_hdr == IP_PROTO_TCP) {
            hdr.swtraces.setValid();
            hdr.swtraces.qtimestamp = (timestamp_t)standard_metadata.enq_timestamp;
            hdr.swtraces.qdepth = (qdepth_t)standard_metadata.deq_qdepth;
        }
```
Wireshrak seems to have no way to capture mininet traffic information in docker.
I want to try mn-pcap to see if I can capture some traffic, so maybe I can figure out what's wrong.
I have not used this mn-pcap, can you give some advice?
Cheers

**Comment by ederollora** [comment]
I always used INT post TCP or UDP for the reason mentioned.

Some ideas about the possible problem:

- is the packet being sent to h1b by last hop switch or other switches? Are you properly forwarding packets?
- Checksum in ipv6 or tcp might be wrong or non existent. Make sure P4 program is not discarding them. Mininet might work just fine with wrong checksums but you never know.
- Try to use udp instead of tcp. Then make a server with netcat and check if packet arrives. Then you can narrow down if the packet is dropped at a partilcular layer.
- please sniff packets and let us know
 Thanks

**Comment by La-Joya** [comment]
Thank you for taking the time out of your busy schedule to answer my question.
Your suggestion is very meaningful, I will experiment one by one to troubleshoot this problem.
If I can do it, I will definitely let you know.
Thanks.
